File: Rishabh_Gupta/RemoveDuplicates.java
class Node {
    int val;
    Node next;

    Node(int val) {
        this.val = val;
        this.next = null;
    }
} 
//Remove Duplicate Element in a Linked List

public class RemoveDuplicates {

    public Node removeDuplicates(Node head) {
        Node current = head;
        while (current != null && current.next != null) {
            if (current.val == current.next.val) {
                current.next = current.next.next;
            } else {
                current = current.next;
            }
        }
        return head;
    }

    public static void printList(Node head) {
        Node current = head;
        while (current != null) {
            System.out.print(current.val + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        Node node5 = new Node(5);
        Node node4_3 = new Node(4);
        Node node4_2 = new Node(4);
        Node node4_1 = new Node(4);
        Node node2_2 = new Node(2);
        Node node2_1 = new Node(2);
        Node head = new Node(1);

        head.next = node2_1;
        node2_1.next = node2_2;
        node2_2.next = node4_1;
        node4_1.next = node4_2;
        node4_2.next = node4_3;
        node4_3.next = node5;

        System.out.println("Linked list before removing duplicates:");
        printList(head);

        RemoveDuplicates solution = new RemoveDuplicates();
        Node newHead = solution.removeDuplicates(head);

        System.out.println("\nLinked list after removing duplicates:");
        printList(newHead);
    }
}

File: Aditya Vardhan Singh/Code Satra Problems/Day 6/C/RemoveDuplicates.c
/*** Custom linked list header file ***/
#include "./List.h"

/*** Procedure to remove duplicate elements form a list ***/
void unique(List ** head) {
    // only a single element or none
    // in that case, list already has unique element
    if (!*head || !(*head)->next) return;

    List *ptr = *head;
    List *temp;
    while (ptr->next) {
        if (ptr->data == ptr->next->data) { 
            // duplicates found at ptr and ptr->next
            temp = ptr->next;
            ptr->next = ptr->next->next;
            free(temp);
        } else {
            ptr = ptr->next;
        }
    }
}

/*** Driver code ***/
int main(int argc, const char **argv) {
    // Check file arguments
    if (argc != 2) {
        printf("Usage: %s <input_file>\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    // Open file and scan list
    FILE *fin = fopen(argv[1], "r");
    List* head = NULL;
    int data, n;
    fscanf(fin, "%d", &n);
    while (n--) {
        fscanf(fin, "%d", &data);
        push_back(&head, data);
    }
    // File operations end

    // Show list
    printf("Original list:\n");
    print_list(&head);

    // Remove duplicates from list
    unique(&head);

    // Show list
    printf("List after removing duplicates:\n");
    print_list(&head);

    // Close and exit drive code
    fclose(fin);
    return EXIT_SUCCESS;
}